library/x509write_csr.c

```diff
@@ -273,24 +280,33 @@ int mbedtls_x509write_csr_pem( mbedtls_x509write_csr *ctx, unsigned char *buf, s
                        int (*f_rng)(void *, unsigned char *, size_t),
                        void *p_rng )
 {
-    int ret;
-    unsigned char output_buf[4096];
+    int ret = 0;
+    unsigned char *output_buf;
+    enum { OUTPUT_BUF_LEN = 4096 };
     size_t olen = 0;
 
-    if( ( ret = mbedtls_x509write_csr_der( ctx, output_buf, sizeof(output_buf),
+    output_buf = mbedtls_calloc(1, OUTPUT_BUF_LEN);
+    if (output_buf == NULL)
+    {
+        return MBEDTLS_ERR_X509_ALLOC_FAILED;
+    }
+
+    if( ( ret = mbedtls_x509write_csr_der( ctx, output_buf, OUTPUT_BUF_LEN,
                                    f_rng, p_rng ) ) < 0 )
     {
-        return( ret );
+        goto cleanup;
     }
 
     if( ( ret = mbedtls_pem_write_buffer( PEM_BEGIN_CSR, PEM_END_CSR,
-                                  output_buf + sizeof(output_buf) - ret,
+                                  output_buf + OUTPUT_BUF_LEN - ret,
                                   ret, buf, size, &olen ) ) != 0 )
     {
-        return( ret );
+        goto cleanup;
     }
 
-    return( 0 );
+cleanup:
```

Comment:
Conventionally we use `exit` as the label name for cleanup-and-exit.

Reply:
Within the X.509 module, `cleanup` seemed to be about as common.

Reply:
Indeed, seems we prefer `cleanup` in crypto, `exit` in TLS and in tests, and X509 is in between.